Module not found: Can't resolve 'fs' when using with Next.js (Turbopack/Webpack)

# `Module not found: Can't resolve 'fs'` when using with Next.js (Turbopack/Webpack)

## Summary

The library fails to load in browser environments when used with Next.js (both Turbopack and Webpack) due to hardcoded `require('fs')` and `require('path')` calls in the bundled distribution files.

## Environment

| Component | Version |
|-----------|---------|
| maplibre-gl-lidar | latest |
| Next.js | 16.1.1 |
| Bundler | Turbopack (also tested with Webpack) |
| Node.js | 20.x |
| OS | Windows 11 |

## Steps to Reproduce

1. Create a Next.js project
2. Install `maplibre-gl-lidar`:
   ```bash
   npm install maplibre-gl-lidar
   ```
3. Import the library in a client component:
   ```tsx
   'use client';
   import { LidarControl } from 'maplibre-gl-lidar';
   ```
4. Run the development server:
   ```bash
   npm run dev
   ```

## Expected Behavior

The library should load successfully in the browser without requiring Node.js-specific modules.

## Actual Behavior

Build fails with the following error:

```
Module not found: Can't resolve 'fs'
  6664 |       }
  6665 |       if (!nodeFS)
> 6666 |         nodeFS = require("fs");
       |                  ^^^^^^^^^^^^^
  6667 |       if (!nodePath)
  6668 |         nodePath = require("path");

Import trace:
  ./node_modules/maplibre-gl-lidar/dist/LidarLayerAdapter-kyyfQw05.js
  ./node_modules/maplibre-gl-lidar/dist/index.mjs
```

## Root Cause Analysis

The bundled file `LidarLayerAdapter-kyyfQw05.js` contains direct `require('fs')` and `require('path')` calls (lines 6666-6668). These are Node.js-only modules that don't exist in browser environments.

This appears to be coming from a dependency (possibly related to COPC/LAS file parsing) that wasn't properly configured for browser bundling.

## Attempted Workarounds

### 1. Dynamic import with `ssr: false`
Does not work because Turbopack still analyzes the module at build time.

```tsx
const LidarControl = dynamic(() => import('maplibre-gl-lidar'), { ssr: false });
```

### 2. Webpack fallback configuration
Partially works with Webpack but not with Turbopack:

```js
// next.config.js
webpack: (config, { isServer }) => {
  if (!isServer) {
    config.resolve.fallback = { fs: false, path: false };
  }
  return config;
}
```

### 3. Turbopack resolveAlias
Not effective for this case as the `require()` calls are hardcoded in the bundle.

## Suggested Fix

The library's build configuration should be updated to:

1. **Mark `fs` and `path` as external/browser-excluded dependencies** during the build process
2. **Use browser-compatible alternatives** for any file system operations (if needed)
3. **Conditionally load Node.js-specific code paths** only when running in Node.js environments

## Additional Context

This library is designed for browser-based MapLibre GL visualization of COPC point clouds. It should not require Node.js filesystem modules at runtime in the browser.

--- a/vite.config.ts
+++ b/vite.config.ts
@@ -5,6 +5,18 @@ import dts from 'vite-plugin-dts';
 
 export default defineConfig({
   plugins: [
+    // Plugin to redirect @loaders.gl/las bundled laz-perf to browser version
+    // This avoids Node.js-specific require('fs') and require('path') calls
+    {
+      name: 'redirect-loaders-gl-laz-perf',
+      enforce: 'pre',
+      resolveId(id, importer) {
+        if (id.includes('laz-perf') && importer?.includes('@loaders.gl/las')) {
+          return resolve(__dirname, 'node_modules/laz-perf/lib/web/laz-perf.js');
+        }
+        return null;
+      },
+    },
     react(),
     dts({
       include: ['src'],
@@ -15,6 +27,8 @@ export default defineConfig({
   resolve: {
     alias: {
       '@': resolve(__dirname, 'src'),
+      // Force laz-perf to use browser version (no fs/path requires)
+      'laz-perf': resolve(__dirname, 'node_modules/laz-perf/lib/web/index.js'),
     },
   },
   assetsInclude: ['**/*.wasm'],

